src/plugin_output_processing/provider.py

```diff
@@ -18,60 +18,77 @@
 
 import ollama
 import openai
+from httpx import ConnectError
 from loguru import logger
 
+OLLAMA_NAME = "ollama"
+OPENAI_NAME = "openai"
+
 
 class ProviderError(Exception):
     pass
 
 
 class Provider:
-
     name: str
-    models: list[str] = []
-    available: bool
+    available: bool = False
     url: str = None
-
-    def get_model(self, model) -> str:
-        if not model:
-            logger.warning(f"Model not set for {self.name}. Using default.")
-            model = self.default
-        if model not in self.models:
-            logger.warning(f"Model {model} not found in {self.name} models.")
-            model = self.default
+    model: str
+
+    def get_model(
+        self,
+        list_models: list[str],
+        model: str | None = None,
+        default_model: str | None = None,
+    ) -> str:
+        if default_model is None:
+            default_model = list_models[0]
+        if model is None:
+            return default_model
+        if model not in list_models:
+            logger.warning(f"{model} not found in model list, using default instead.")
+            return default_model
         return model
 
 
 class Ollama(Provider):
-
-    name = "ollama"
+    name = OLLAMA_NAME
     url = f"http://{os.environ.get('OLLAMA_HOST', 'localhost')}:11434"
 
-    def __init__(self) -> None:
-        self.models = self._list_models()
-        self.default = self.models[0] if self.models else None
-        self.available = self.models != []
+    def __init__(self, model: str = None) -> None:
+        self.model = self.fetch_model(model)
+        self.available = self.model is not None
 
-    def _list_models(self) -> list[str]:
+    def fetch_model(self, model: str | None = None) -> str | None:
         try:
-            return [model["name"] for model in ollama.list()["models"]]
-        except Exception:
-            logger.debug(f"Could not list models for {self.name}.")
-            return []
+            models = ollama.list()["models"]
+        except ConnectError:
+            logger.debug(f"{self.name} is not available.")
+            return None
+        if len(models) == 0:
+            small_model = "qwen2:0.5b"
+            logger.warning(
+                f"{self.name} can be reached but no models found, downloading {small_model}."
+            )
+            logger.info("Beware, this will take some time.")
+            ollama.pull(small_model)
+            return small_model
+        list_models = [m["name"] for m in models]
+        return self.get_model(list_models, model)
```

Comment:
```suggestion
class Ollama(Provider):
    name = OLLAMA_NAME
    url = f"http://{os.environ.get('OLLAMA_HOST', 'localhost')}:11434"

    def fetch_model(self, model: str | None = None) -> str | None:
        try:
            self.models = [model["name"] for model in ollama.list()["models"]]
        except ConnectError:
            logger.debug(f"{self.name} is not available.")
            return None
        if not self.models:
            small_model = "qwen2:0.5b"
            logger.warning(
                f"{self.name} can be reached but no models found, downloading {small_model}."
            )
            logger.info("Beware, this will take some time.")
            ollama.pull(small_model)
            self.models.append(small_model)

        return self.get_model(model)
```

Few changes:

- using attributes `self.default` and `self.models`
- Get list of models names in one code line.
- Add name of `small_model` in the `self.models` list.

Reply:
* Using attributes was done in a previous commit
* OK for the one liner
* I prefer returning small_model directly (because we pulled it, we know the name exists) rather than going to another method. I know the time used in `self.get_model` is imperceptible, but if I can avoid doing operation, I'll do it. I don't think it complexify the readibility of the code.